### PR TITLE
Advancement

### PR DESCRIPTION
**Pull request by nicolasaunai**
#116

### CODE REVIEW

**Comment by nicolasaunai** [comment]
at this point test Ohm failure is normal because I have removed some terms in the equation in cpp and not in python test

File: src/diagnostic/diagnostic_manager.h [hunk 1]
```diff
@@ -147,22 +147,21 @@ void DiagnosticsManager<Writer>::dump(double timeStamp, double timeStep)
...
 
     for (auto const* diag : activeDiagnostics)
     {
-        lastWrite_[diag->type + diag->quantity]++;
+        nextWrite_[diag->type + diag->quantity]++;
     }
-}
-
-
+} // namespace PHARE::diagnostic
```
Comment (Dekken) [inline src/diagnostic/diagnostic_manager.h:164]: :|

File: src/amr/messengers/hybrid_hybrid_messenger_strategy.h [hunk 2]
```diff
@@ -449,6 +465,7 @@ namespace amr
 
             magneticGhosts_.fill(hybridModel.state.electromag.B, levelNumber, initDataTime);
             electricGhosts_.fill(hybridModel.state.electromag.E, levelNumber, initDataTime);
+            // currentGhosts_.fill(hybridModel.state.J, levelNumber, initDataTime); //TODO?
```
Comment (Dekken) [inline src/amr/messengers/hybrid_hybrid_messenger_strategy.h:468]: is there an issue for this TODO?

File: src/core/numerics/ohm/ohm.h [hunk 3]
```diff
@@ -321,9 +321,9 @@ namespace core
             for (auto ix = ix0; ix <= ix1; ++ix)
             {
                 Ey(ix) = ideal1D_(Ve, B, {ix}, ComponentTag<Component::Y>{})
-                         + pressure_(n, Pe, {ix}, ComponentTag<Component::Y>{})
-                         + resistive_(J, {ix}, ComponentTag<Component::Y>{})
-                         + hyperresistive_(J, {ix}, ComponentTag<Component::Y>{});
+                         + 0 * pressure_(n, Pe, {ix}, ComponentTag<Component::Y>{})
```
Comment (Dekken) [inline src/core/numerics/ohm/ohm.h:324]: please explain

File: src/core/numerics/pusher/pusher.h [hunk 4]
```diff
@@ -3,19 +3,23 @@
...
-             typename ParticleSelector, typename BoundaryCondition, typename GridLayout>
+             typename BoundaryCondition, typename GridLayout>
     class Pusher
     {
     protected:
-        using ParticleRange = Range<ParticleIterator>;
-
+        using ParticleRange             = Range<ParticleIterator>;
+        static auto constexpr dimension = GridLayout::dimension;
+        using ParticleSelector          = std::function<bool(Particle<dimension> const&)>;
```
Comment (Dekken) [inline src/core/numerics/pusher/pusher.h:22]: is the class redundant now?

File: src/phare/phare_init.py [hunk 5]
```diff
@@ -1,41 +1,117 @@
...
     boundary_types="periodic", # boundary condition, string or tuple, length == len(cell) == len(dl)
-    cells=65,                  # integer or tuple length == dimension
-    dl=1./65,                  # mesh size of the root level, float or tuple
-    max_nbr_levels=2,          # (default=1) max nbr of levels in the AMR hierarchy
-    refinement_boxes = {"L0":{"B0":[(10,),(50,)]}},
-    diag_options = {"format":"phareh5", "options": {"dir": "phare_outputs"}}
+    cells=100,                # integer or tuple length == dimension
+    dl=0.3,                  # mesh size of the root level, float or tuple
+    max_nbr_levels=1,          # (default=1) max nbr of levels in the AMR hierarchy
+    #refinement_boxes={"L0": {"B0": [(10, ), (20, )]}},
```
Comment (Dekken) [inline src/phare/phare_init.py:22]: ?

File: src/python3/data_wrangler.cpp [hunk 6]
```diff
@@ -70,12 +70,13 @@ class PatchLevel
...
-    using GridLayout  = typename HybridModel::gridLayout_type;
+    using GridLayout  = typename HybridModel::gridlayout_type;
 
     PatchLevel(amr::Hierarchy& hierarchy, HybridModel& model, size_t lvl)
-        : hierarchy_{hierarchy}
+        : lvl_(lvl)
+        , hierarchy_{hierarchy}
         , model_{model}
-        , lvl_(lvl)
+
```
Comment (Dekken) [inline src/python3/data_wrangler.cpp:79]: rm

File: src/simulator/simulator.h [hunk 7]
```diff
@@ -134,16 +133,23 @@ class Simulator : public ISimulator
 
 
     double startTime() override { return 0.; }
+
     double endTime() override { return finalTime_; }
     double timeStep() override { return dt_; }
     double currentTime() override { return currentTime_; }
 
-    void advance() override
+    void advance(double dt) override
```
Comment (Dekken) [inline src/simulator/simulator.h:141]: this is another function directly exposed to python, so could do with a try/catch(print)

File: src/solver/multiphysics_integrator.h [hunk 8]
```diff
@@ -336,7 +336,7 @@ namespace solver
         virtual double getMaxFinerLevelDt(const int finerLevelNumber, const double coarseDt,
                                           const SAMRAI::hier::IntVector& ratio) override
         {
-            return std::pow(coarseDt, ratio.max());
+            return coarseDt / (ratio.max() * ratio.max()) / 2.;
```
Comment (Dekken) [inline src/solver/multiphysics_integrator.h:339]: no changes requested, just adding for posterity:
https://stackoverflow.com/questions/18165047/multiply-by-0-5-rather-than-divide-by-2

File: src/solver/solvers/solver_ppc.h [hunk 9]
```diff
@@ -132,151 +182,335 @@ namespace solver
...
-        IPhysicalModel<AMR_Types>& model,
-        amr::IMessenger<IPhysicalModel<AMR_Types>>& fromCoarserMessenger, const double currentTime,
-        const double newTime)
+    void
+    SolverPPC<HybridModel, AMR_Types>::advanceLevel(std::shared_ptr<hierarchy_t> const& hierarchy,
+                                                    int const levelNumber, IPhysicalModel_t& model,
+                                                    IMessenger& fromCoarserMessenger,
+                                                    const double currentTime, const double newTime)
     {
         // bool constexpr withTemporal{true};
```
Comment (Dekken) [inline src/solver/solvers/solver_ppc.h:191]: still needed?

File: src/solver/solvers/solver_ppc.h [hunk 10]
```diff
@@ -132,151 +182,335 @@ namespace solver
...
-        //  - coarse to fine ghost region
-        // accumulate those that are within the domain after being pushed (before pivot)
-        // see particle design code
 
-        // now some of the nodes close to the boundary of the patches (and level)
-        // are incomplete because they may have recieved contributions from particles outside
-        // their domain that would have entered the purple region during the push
-        // therefore we need to re-fill the purple region and accumulate that density
-        // this is done by calling a messenger to fill the moments.
+        // return newTime;
```
Comment (Dekken) [inline src/solver/solvers/solver_ppc.h:221]: will this be returned in future?

File: src/solver/solvers/solver_ppc.h [hunk 11]
```diff
@@ -132,151 +182,335 @@ namespace solver
...
 
-        auto coarserLevel = hierarchy->getPatchLevel(levelNumber - 1);
-        fromCoarser.fillIonMomentGhosts(hybridState.ions, *level, currentTime, newTime);
 
 
 
+    template<typename HybridModel, typename AMR_Types>
+    void SolverPPC<HybridModel, AMR_Types>::predictor1_(level_t& level, HybridModel& model,
+                                                        Messenger& fromCoarser,
+                                                        const double currentTime,
```
Comment (Dekken) [inline src/solver/solvers/solver_ppc.h:230]: cosmetic: do we prefer "double const" ?

File: src/solver/solvers/solver_ppc.h [hunk 12]
```diff
@@ -132,151 +182,335 @@ namespace solver
...
+        for (auto& patch : level)
+        {
+            auto _      = rm.setOnPatch(*patch, electromag, ions);
+            auto layout = PHARE::amr::layoutFromPatch<GridLayout>(*patch);
+            ionUpdater_.updateIons(ions, layout);
+
+            // no need to update time, since it has been done before
+        }
+    }
+} // namespace solver
```
Comment (Dekken) [inline src/solver/solvers/solver_ppc.h:513]: cosmetic:

less indentation with

```
namespace PHARE::solver {
```

if you feel like it

File: src/core/numerics/ampere/ampere.h [hunk 13]
```diff
@@ -192,7 +192,7 @@ namespace core
                 throw std::runtime_error(
                     "Error - Ampere - GridLayout not set, cannot proceed to calculate ampere()");
             }
-            std::cout << "I'm solving ampere " << GridLayout::dimension << "\n";
+            // std::cout << "I'm solving ampere " << GridLayout::dimension << "\n";
```
Comment (Dekken) [inline src/core/numerics/ampere/ampere.h:195]: rm?

File: src/amr/messengers/hybrid_hybrid_messenger_strategy.h (hunk 2 above)
Comment (nicolasaunai) [reply src/amr/messengers/hybrid_hybrid_messenger_strategy.h:468]: nope. I think it was a question for myself whether or not to do this here.
actually, this function is called here 

https://github.com/nicolasaunai/PHARE/blob/advancement/src/solver/level_initializer/hybrid_level_initializer.h#L60

i.e. just after the model initializes the root level.
However the model only loads the particles in each pop and initializes the magnetic field:

https://github.com/nicolasaunai/PHARE/blob/advancement/src/solver/physical_models/hybrid_model.h#L72-L79

so when the fillRootGhosts() is called, the current is actually 0, thus filling ghosts is useless.

The current J is now calculated in the hybrid level initializer here

https://github.com/nicolasaunai/PHARE/blob/advancement/src/solver/level_initializer/hybrid_level_initializer.h#L102-L116

mainly to be used by Ohm to get the electric field initialized on root level


long story short : I have deleted that line.

File: src/core/numerics/ohm/ohm.h (hunk 3 above)
Comment (nicolasaunai) [reply src/core/numerics/ohm/ohm.h:324]: the ohm's law is : 

E = -V_e x B - grad(P_e)/(e*n_e) + eta*j - nu*laplacian(J).

-V_e x B is called the "ideal term"
-grad(Pe)/(e*ne) is the "pressure term"
eta*j is the resistive term
-n*laplacian(J) is the hyper resistive term.

as a first step to see whether PHARE works, it's simpler to have only the ideal term in ohm's law, thus zeroing the others.

They will reappear step by step in the (very) near future, but they are tougher to handle.

Comment (Dekken) [reply src/core/numerics/ohm/ohm.h:324]: sorry I should have been more specific, I was referring to the "0 *"

File: src/core/numerics/pusher/pusher.h (hunk 4 above)
Comment (nicolasaunai) [reply src/core/numerics/pusher/pusher.h:22]: ah yeah, I wanted to remove it and forgot.
Yeah, so the reason behind the lambdas is that:

- first, it's simpler to define a lambda than a functor
- second it's also more adequate because you'll notice that the class ParticleSelector basically assume you're going to select from a box (or a collection of, as long as the criteria is "isIn"). But now in the IonUpdater, when we push patchGhost particles, we want a selector that checks whether they leave the patch ghost area, which is basically "in ghost box but not in domain box".... ParticleSelector didn't work for that, but lambdas can do whatever you want.... so simpler.

I'll remove the ParticleSelector class

File: src/phare/phare_init.py (hunk 5 above)
Comment (nicolasaunai) [reply src/phare/phare_init.py:22]: indeed :-)
so, we go step by step. For now this PR is about having PHARE working with only a coarse level. I've tested 1 coarse level with a single or 4 patches with the Alfvén wave test as you've seen.

next step will be to have multiple (fixed) refinement level, but we need to adress #181  before.

that's why have not removed but commented the line. It's going to reappear in the next PR.

and the step after will consist in having a dynamic refinement criterion (or criteria) i.e. cell tagging.

File: src/python3/data_wrangler.cpp (hunk 6 above)
Comment (nicolasaunai) [reply src/python3/data_wrangler.cpp:79]: done

File: src/simulator/simulator.h (hunk 7 above)
Comment (nicolasaunai) [reply src/simulator/simulator.h:141]: ??

Comment (Dekken) [reply src/simulator/simulator.h:141]: similarly to https://github.com/PHAREHUB/PHARE/blob/a5eb3732e486cf5d7b7286e3a9b3f1a21f1122bb/src/simulator/simulator.h#L112

File: src/phare/phare_init.py (hunk 5 above)
Comment (Dekken) [reply src/phare/phare_init.py:22]: ok thanks!

File: src/core/numerics/ohm/ohm.h (hunk 3 above)
Comment (nicolasaunai) [reply src/core/numerics/ohm/ohm.h:324]: I know you did, but I gave you a bonus explanation

File: src/amr/messengers/hybrid_hybrid_messenger_strategy.h (hunk 2 above)
Comment (Dekken) [reply src/amr/messengers/hybrid_hybrid_messenger_strategy.h:468]: > long story short : I have deleted that line.

lmao

File: src/core/numerics/ohm/ohm.h (hunk 3 above)
Comment (Dekken) [reply src/core/numerics/ohm/ohm.h:324]: > They will reappear step by step in the (very) near future, but they are tougher to handle.

ok this is the important part I think, the "0 *" is temporary

File: src/core/numerics/ampere/ampere.h (hunk 13 above)
Comment (nicolasaunai) [reply src/core/numerics/ampere/ampere.h:195]: done

File: src/diagnostic/diagnostic_manager.h (hunk 1 above)
Comment (nicolasaunai) [reply src/diagnostic/diagnostic_manager.h:164]: done

File: src/simulator/simulator.h (hunk 7 above)
Comment (nicolasaunai) [reply src/simulator/simulator.h:141]: ok done

File: src/solver/solvers/solver_ppc.h (hunk 9 above)
Comment (nicolasaunai) [reply src/solver/solvers/solver_ppc.h:191]: don't even know what that was... removed

File: src/solver/solvers/solver_ppc.h (hunk 10 above)
Comment (nicolasaunai) [reply src/solver/solvers/solver_ppc.h:221]: not sure yet, I've let that, probably assessed in the multiple level PR

File: src/solver/solvers/solver_ppc.h (hunk 11 above)
Comment (nicolasaunai) [reply src/solver/solvers/solver_ppc.h:230]: oh yeah... what was I thinking 🤮 ;-)

File: src/solver/solvers/solver_ppc.h (hunk 12 above)
Comment (nicolasaunai) [reply src/solver/solvers/solver_ppc.h:513]: I felt
